### ISSUE TITLE
Problems installing on cygwin

### ISSUE THREAD

**Issue by mijoharas** [rvm/rvm#3309]
Fresh install of `rvm`:

```
++ 1424284498.795828300 /scripts/functions/utility_logging : __rvm_log_command_simple()  83 > gem install /home/mijoharas/.rvm/gem-cache/gem-wrappers-1.2.7.gem --local --no-ri --no-rdoc
Building native extensions.  This could take a while...
      0 [main] ruby 10148 child_info_fork::abort: address space needed by 'fcntl.so' (0x810000) is already occupied
      0 [main] ruby 1252 child_info_fork::abort: address space needed by 'fcntl.so' (0x810000) is already occupied
      0 [main] ruby 10040 child_info_fork::abort: address space needed by 'fcntl.so' (0x810000) is already occupied
      0 [main] ruby 3664 child_info_fork::abort: address space needed by 'fcntl.so' (0x810000) is already occupied
      0 [main] ruby 5300 child_info_fork::abort: address space needed by 'fcntl.so' (0x810000) is already occupied
      0 [main] ruby 4840 child_info_fork::abort: address space needed by 'openssl.so' (0x22A0000) is already occupied
```

I'm assuming it may be something to do with the requirements to `rebaseall`, ([see here](http://thread.gmane.org/gmane.os.cygwin/126151/focus=126196) for a ruby specific example).

**Comment by mijoharas** [comment]
Installed again with:

```
rvm implode
curl -L https://get.rvm.io | bash -s stable
rvm install ruby --trace
```

to verify, this resulted in:

```
<snip>
GEM_PATH=/home/hauserra/.rvm/gems/ruby-2.2.0@global
+ 11:12:16 /home/hauserra/.rvm/scripts/functions/utility_logging:75 __rvm_log_command_debug:8 printf %b 'command(7): command gem install /home/hauserra/.rvm/gem-cache/gem-wrappers-1.2.7.gem --local --no-ri --no-rdoc\n'
command(7): command gem install /home/hauserra/.rvm/gem-cache/gem-wrappers-1.2.7.gem --local --no-ri --no-rdoc
+ 11:12:16 /home/hauserra/.rvm/scripts/functions/utility_logging:101 __rvm_log_command_internal:14 __rvm_log_dotted '/home/hauserra/.rvm/log/1424343573_ruby-2.2.0/gem.install.gem-wrappers->=1.2.7.log' 'installing gem /home/hauserra/.rvm/gem-cache/gem-wrappers-1.2.7.gem --local --no-ri --no-rdoc' command gem install /home/hauserra/.rvm/gem-cache/gem-wrappers-1.2.7.gem --local --no-ri --no-rdoc
+ 11:12:16 /home/hauserra/.rvm/scripts/functions/utility_logging:183 __rvm_log_dotted:2 typeset __log_file __message __iterator __result __local_rvm_trace_flag
+ 11:12:16 /home/hauserra/.rvm/scripts/functions/utility_logging:184 __rvm_log_dotted:3 __log_file='/home/hauserra/.rvm/log/1424343573_ruby-2.2.0/gem.install.gem-wrappers->=1.2.7.log'
+ 11:12:16 /home/hauserra/.rvm/scripts/functions/utility_logging:185 __rvm_log_dotted:4 __message='installing gem /home/hauserra/.rvm/gem-cache/gem-wrappers-1.2.7.gem --local --no-ri --no-rdoc'
+ 11:12:16 /home/hauserra/.rvm/scripts/functions/utility_logging:186 __rvm_log_dotted:5 shift 2
+ 11:12:16 /home/hauserra/.rvm/scripts/functions/utility_logging:187 __rvm_log_dotted:6 __result=0
+ 11:12:16 /home/hauserra/.rvm/scripts/functions/utility_logging:188 __rvm_log_dotted:7 __local_rvm_trace_flag=1
+ 11:12:16 /home/hauserra/.rvm/scripts/functions/utility_logging:190 __rvm_log_dotted:9 if((  1  ))
+ 11:12:16 /home/hauserra/.rvm/scripts/functions/utility_logging:193 __rvm_log_dotted:12 then curshset -x
+ 11:12:16 /home/hauserra/.rvm/scripts/functions/utility_logging:194 __rvm_log_dotted:13 then curshgem install /home/hauserra/.rvm/gem-cache/gem-wrappers-1.2.7.gem --local --no-ri --no-rdoc
+ 11:12:16 /home/hauserra/.rvm/scripts/functions/utility_logging:194 __rvm_log_dotted:13 then cursh pipetee -a '/home/hauserra/.rvm/log/1424343573_ruby-2.2.0/gem.install.gem-wrappers->=1.2.7.log'
Building native extensions.  This could take a while...
      0 [main] ruby 12920 child_info_fork::abort: address space needed by 'fcntl.so' (0x350000) is already occupied
      0 [main] ruby 7896 child_info_fork::abort: address space needed by 'fcntl.so' (0x350000) is already occupied
      0 [main] ruby 9148 child_info_fork::abort: address space needed by 'fcntl.so' (0x350000) is already occupied
      0 [main] ruby 11144 child_info_fork::abort: address space needed by 'fcntl.so' (0x350000) is already occupied
      0 [main] ruby 13004 child_info_fork::abort: address space needed by 'fcntl.so' (0x350000) is already occupied
      0 [main] ruby 12308 child_info_fork::abort: address space needed by 'fcntl.so' (0x350000) is already occupied
      0 [main] ruby 4232 child_info_fork::abort: address space needed by 'fcntl.so' (0x350000) is already occupied
      0 [main] ruby 10800 child_info_fork::abort: address space needed by 'fcntl.so' (0x350000) is already occupied
<snip>
```
with the child_info_fork::abort commands carrying on (sometimes with different *.so files).

**Comment by mijoharas** [comment]
Can also repro more minimally with `run_gem_wrappers_regenerate` I can post output for that if needed.

**Comment by mijoharas** [comment]
I ended up digging pretty deep into this today. How does rvm handle `rebaseall`? I can get a working ruby installation both by installing from source, then:

```
cd my_ruby_install_dir
find `pwd` -iname '*.so' > /tmp/ruby_so.tmp
# in dash shell with all cygwin processes closed
rebaseall -v -T /tmp/ruby_so.tmp
```

or by using `ruby-build` as an rvm plugin and then doing:

```
cd ~/.rbenv/versions/$my_ruby_version
find `pwd` -iname '*.so' > /tmp/ruby_so.tmp
# in dash shell with all cygwin processes closed
rebaseall -v -T /tmp/ruby_so.tmp
```

this is because any .so files that are installed need to be set to specific positions in memory so that cygwin's forking process doesn't die.

the relevant parts of rebaseall that searches for the files are:

```
DefaultSuffixes='dll|so|oct'
#<snip>
find /etc/setup -name '*.lst.gz' | xargs gzip -d -c |
      grep -E "\.($Suffixes)\$" |
      sed -e '/\/cygwin1\.dll$/d' -e '/\/cyglsa.*\.dll$/d' \
          -e '/sys-root\/mingw/d' -e 's/^/\//' \
          -e '/\/d\?ash\.exe$/d' -e '/\/rebase\.exe$/d' >"${TmpFile}"
    # some interpreters include a method for installing addons outside of the
    # package manager, such as CPAN and RubyGems.
    for d in /usr/lib/perl5/site_perl /usr/lib/py*/site-packages \
             /usr/lib/php /usr/lib/R/site-library /usr/lib/rub*/gems \
             /usr/lib/octave/site
    do
      if [ -d $d ]
      then
        find $d -type f | grep -E "\.($Suffixes)\$" >>"${TmpFile}"
      fi
    done
```
Now it seems like we could simplify the rebaseall process by adding `/home/*/.rvm` (and for the rbenv case `/home/*/.rbenv/versions/*` to the for loop above, but we still have the problem that we can't call rebaseall with cygwin processes open and running. I'm not sure how we can get around that in an automated, non-horrifically hacky way.

**Comment by mpapis** [comment]
rvm runs in the cygwin process,we would have to print a big warning with commands to run outside of the process ... btw. do you know why this happens? many people before did not had this problem.

**Comment by mijoharas** [comment]
I'm wondering that myself, looks like there was [this issue here](https://github.com/wayneeseguin/rvm/issues/2952) but nothing before that. I've never had much need to use a windows system much so I only noticed it recently.

**Comment by mijoharas** [comment]
(as in I've never successfully used rvm on cygwin before now, works great on linux though, thanks :smile: )

**Comment by mpapis** [comment]
maybe start with opening a bug for ruby, they might know what is going on

**Comment by mijoharas** [comment]
I'm not sure if it is a ruby bug, AFAIK it's expected behaviour with cygwin.

**Comment by parkerjm** [comment]
I am also experiencing this issue. rebaseall from dash did not help.

**Comment by mpapis** [comment]
it's a compilation error, please report it for ruby they are best skilled to analyse it and to find a fix for it

**Comment by havenwood** [comment]
Closing since some Cygwin compilation errors have been resolved upstream with the release of Ruby 2.1.7 and Ruby 2.2.3. Please open a new Issue for any compilation errors that still occur in 2.1.7+ or 2.2.3+.
